### PR TITLE
Export condensed distance matrix to a pandas object #1397

### PR DESCRIPTION
**Pull request by virginiasaulnier**
I've implemented to_series to create a Pandas series for DissimilarityMatrix in stats.distance _base.py.  I've also created 3 tests in test_base.py.  All tests pass.

### CODE REVIEW

File: skbio/stats/distance/_base.py [hunk 1]
```diff
@@ -487,6 +487,46 @@ def to_data_frame(self):
...
+             a    b    c
+        a  0.0  1.0  2.0
+        b  1.0  0.0  3.0
+        c  2.0  3.0  0.0
+<<<<<<< HEAD
+        
+=======
+
+>>>>>>> 49864617eadb042c3f5c319c2a30694bb483e3f0
+        >>> df.to_series()
```
Comment (mortonjt) [inline skbio/stats/distance/_base.py]: Should be dm.to_series() right?  Also don't forget the to resolve merge conflicts.

File: skbio/stats/distance/_base.sublime-workspace [hunk 2]
```diff
@@ -0,0 +1,169 @@
+{
```
Comment (mortonjt) [inline skbio/stats/distance/_base.sublime-workspace]: Don't forget to delete this.

**Comment by mortonjt** [comment]
Just realized that this may not be ready to review.  Let us know when you are ready.  Thanks!

**Comment by virginiasaulnier** [comment]
Jamie,

The build passed.  Should be ready for review.

Thanks,
Virginia

On Jul 27, 2016, at 4:34 PM, Jamie Morton <notifications@github.com<mailto:notifications@github.com>> wrote:


Just realized that this may not be ready to review. Let us know when you are ready. Thanks!

File: skbio/stats/distance/_base.sublime-workspace (hunk 2 above)
Comment (mortonjt) [inline skbio/stats/distance/_base.sublime-workspace]: Don't forget this.  We don't want project files accidentally committed in the main repository.

**Comment by ebolyen** [comment]
Thanks for working on this @virginiasaulnier!

Fixes #1397 

As a heads up, we will want to squash these commits, but that can be left until this is ready for merge, that way you don't have to do it a bunch of times.

File: skbio/stats/distance/_base.py [hunk 3]
```diff
@@ -487,6 +487,36 @@ def to_data_frame(self):
...
+           c    3.0
+        c  a    2.0
+           b    3.0
+           c    0.0
+        dtype: float64
+
+        """
+        df = pd.DataFrame(data=self.data, index=self.ids, columns=self.ids)
+        stacked = df.stack()
+        return pd.Series(stacked)
```
Comment (ebolyen) [inline skbio/stats/distance/_base.py]: my understanding is that `stacked` will already be a `pd.Series` was there a specific reason it is passed through it's constructor again?

File: skbio/stats/distance/tests/test_base.py [hunk 4]
```diff
@@ -467,6 +468,29 @@ def test_to_data_frame_default_ids(self):
                            index=['0', '1'], columns=['0', '1'])
         assert_data_frame_almost_equal(df, exp)
 
+    def test_to_series_1x1(self):
+        df = self.dm_1x1.to_series()
```
Comment (ebolyen) [inline skbio/stats/distance/tests/test_base.py]: nitpicky but `df` is probably the wrong variable name here. Try just `series`.

File: skbio/stats/distance/_base.py [hunk 5]
```diff
@@ -487,6 +487,36 @@ def to_data_frame(self):
         """
         return pd.DataFrame(data=self.data, index=self.ids, columns=self.ids)
 
+    @experimental(as_of="0.5.0")
```
Comment (ebolyen) [inline skbio/stats/distance/_base.py]: This should be `"0.5.0-dev"` (we change them to the right version just before release)

**Comment by ebolyen** [comment]
This looks really good @virginiasaulnier! Couple of minor comments, and we will need to add a changelog note under `Features` describing this change (that can be found in the root of the repo).

**Comment by virginiasaulnier** [comment]
Evan,

I’m not sure about this.  I couldn’t pass the df directly into pd.Series, because it would only accept a 1D object.  You may have guessed from the number of commits this is my first attempt at contribution.  I’d be happy to adjust it if you have suggestions.

Thanks,
Virginia
On Jul 28, 2016, at 11:11 AM, Evan Bolyen <notifications@github.com<mailto:notifications@github.com>> wrote:


This looks really good @virginiasaulnier<https://github.com/virginiasaulnier>! Couple of minor comments, and we will need to add a changelog note under Features describing this change (that can be found in the root of the repo).

**Comment by ebolyen** [comment]
@virginiasaulnier No worries, the comment was actually that you already had a series after calling `stack`, so passing the result into `pd.Series` again is unnecessary. Your use of `stack` is already correct.

**Comment by virginiasaulnier** [comment]
So the final line should just say return stacked?


On Jul 28, 2016, at 2:44 PM, Evan Bolyen <notifications@github.com<mailto:notifications@github.com>> wrote:


@virginiasaulnier<https://github.com/virginiasaulnier> No worries, the comment was actually that you already had a series after calling stack, so passing the result into pd.Series again is unnecessary. Your use of stack is already correct.

**Comment by ebolyen** [comment]
I believe so.

**Comment by ebolyen** [comment]
Thanks @virginiasaulnier, now all that is missing is a changelog, please see my comment above. Once that is done we will want to collapse all of these commits into a single one with a descriptive name (see our commit history for examples). To accomplish that you will want to read this resource: https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History

**Comment by jairideout** [comment]
Thanks for working on this @virginiasaulnier! There's some uncertainty on what the output should be, see my comment [here](https://github.com/biocore/scikit-bio/issues/1397#issuecomment-236731563). Let's hold off on merging this until we sort that out.

**Comment by virginiasaulnier** [comment]
@jairideout I updated the code to create the series without 0 values.  I'm having a hard time squashing all the commits though.

**Comment by jairideout** [comment]
Thanks @virginiasaulnier, I'll take a pass through your pull request now. Don't worry about squashing yet as there may be more comments to address. We can help you with squashing when this is ready for merge, or we can merge and squash manually for you.

File: skbio/stats/distance/_base.py [hunk 6]
```diff
@@ -487,6 +487,34 @@ def to_data_frame(self):
         """
         return pd.DataFrame(data=self.data, index=self.ids, columns=self.ids)
 
+    @experimental(as_of="0.5.0-dev")
+    def to_series(self):
```
Comment (jairideout) [inline skbio/stats/distance/_base.py]: Now that `to_series` returns the condensed form distances (i.e. only one triangle, excluding the diagonal), this method should be moved to `DistanceMatrix` since a scikit-bio distance matrix is guaranteed to be symmetric and hollow. `DissimilarityMatrix` can be non-hollow and/or asymmetric so condensed form isn't applicable here.

File: skbio/stats/distance/_base.py [hunk 7]
```diff
@@ -487,6 +487,34 @@ def to_data_frame(self):
...
+        """Create a ``pandas.Series`` from this ``DissimilarityMatrix``.
```
Comment (jairideout) [inline skbio/stats/distance/_base.py]: Change this line to reference `DistanceMatrix`.

Can you add the following description below this summary line?

```
The series will contain distances in condensed form: only distances from one matrix triangle are included,
and the diagonal is excluded. The series' index will be a ``pd.MultiIndex`` relating pairs of IDs to distances.
The pairs of IDs will be in row-major order with respect to the upper matrix triangle.

To obtain all distances (i.e. both upper and lower matrix triangles and the diagonal), use
``DistanceMatrix.to_data_frame``. To obtain *only* the distances in condensed form
(e.g. for use with SciPy), use ``DistanceMatrix.condensed_form``.
```

File: skbio/stats/distance/_base.py [hunk 8]
```diff
@@ -487,6 +487,34 @@ def to_data_frame(self):
...
+
+        Returns
+        -------
+        pd.Series
+            ``pd.Series`` with IDs on index.
```
Comment (jairideout) [inline skbio/stats/distance/_base.py]: Change to:

```
``pd.Series`` with pairs of IDs on the index.
```

File: skbio/stats/distance/_base.py [hunk 9]
```diff
@@ -487,6 +487,34 @@ def to_data_frame(self):
...
+
```
Comment (jairideout) [inline skbio/stats/distance/_base.py]: Can you add a "See Also" section above "Examples"?

```
See Also
--------
to_data_frame
condensed_form
scipy.spatial.distance.squareform
```

File: skbio/stats/distance/_base.py [hunk 10]
```diff
@@ -487,6 +487,34 @@ def to_data_frame(self):
...
+        Examples
+        --------
+        >>> from skbio import DistanceMatrix
```
Comment (jairideout) [inline skbio/stats/distance/_base.py]: Can you expand this example to use a 4x4 distance matrix? With a 3x3, the iteration order isn't clear for the pairs of IDs put on the index (row-major order with respect to the upper matrix triangle).

```python
>>> from skbio import DistanceMatrix
>>> dm = DistanceMatrix([[0, 1, 2, 3],
>>>                      [1, 0, 4, 5],
>>>                      [2, 4, 0, 6],
>>>                      [3, 5, 6, 0]], ids=['a', 'b', 'c', 'd'])
>>> dm.to_series()
a  b    1.0
   c    2.0
   d    3.0
b  c    4.0
   d    5.0
c  d    6.0
dtype: float64
```

File: skbio/stats/distance/_base.py [hunk 11]
```diff
@@ -487,6 +487,34 @@ def to_data_frame(self):
...
+        b  a    1.0
+           c    3.0
+        c  a    2.0
+           b    3.0
+        dtype: float64
+
+        """
+        df = pd.DataFrame(data=self.data, index=self.ids, columns=self.ids)
+        stacked = df.stack()
+        series = stacked[stacked > 0]
```
Comment (jairideout) [inline skbio/stats/distance/_base.py]: The diagonal is guaranteed to be zero but there can be zeros elsewhere in the matrix so this strategy won't always work as expected. We also only want to include pairwise distances from one of the matrix triangles.

You can use `self.condensed_form()` to obtain a 1-D numpy array in condensed form, and then you can build the `pd.MultiIndex` by generating ID pairs with `itertools.combinations`. `to_series` can be implemented as follows:

```python
distances = self.condensed_form()
# `id_pairs` will not be interpreted as a `pd.MultiIndex` if it is an iterable returned by `itertools.combinations`.
id_pairs = list(itertools.combinations(self.ids, 2))
index = pd.Index(id_pairs, tupleize_cols=True)
return pd.Series(data=distances, index=index, dtype=float)
```

File: CHANGELOG.md [hunk 12]
```diff
@@ -6,6 +6,7 @@
 * `DissimilarityMatrix` now has a new constructor method called `from_iterable`. ([#1343](https://github.com/biocore/scikit-bio/issues/1343)).
 * `DissimilarityMatrix` now allows non-hollow matrices. ([#1343](https://github.com/biocore/scikit-bio/issues/1343)).
 * `DistanceMatrix.from_iterable` now accepts a `validate=True` parameter. ([#1343](https://github.com/biocore/scikit-bio/issues/1343)).
+* 'DissimilarityMatrix' now has a new method called 'to_series' to create a Pandas series. ([#1397](https://github.com/biocore/scikit-bio/issues/1397)).
```
Comment (jairideout) [inline CHANGELOG.md]: Change text to reference `DistanceMatrix`. Please use single backticks around `DistanceMatrix` so it renders like other class names in the changelog.

File: skbio/stats/distance/tests/test_base.py [hunk 13]
```diff
@@ -467,6 +468,32 @@ def test_to_data_frame_default_ids(self):
                            index=['0', '1'], columns=['0', '1'])
         assert_data_frame_almost_equal(df, exp)
 
+    def test_to_series_1x1(self):
```
Comment (jairideout) [inline skbio/stats/distance/tests/test_base.py]: Move these tests into `DistanceMatrixTests`

File: skbio/stats/distance/tests/test_base.py [hunk 14]
```diff
@@ -467,6 +468,32 @@ def test_to_data_frame_default_ids(self):
...
+        series = self.dm_1x1.to_series()
+        exp = pd.Series([0.0],
+                        pd.MultiIndex(levels=[['a'], ['a']],
+                        labels=[[0], [0]]))
+        exp = exp[exp > 0]
```
Comment (jairideout) [inline skbio/stats/distance/tests/test_base.py]: The expected value should be `pd.Series([], index=[])` for a 1x1 matrix. I think you're obtaining that on this line but it's better to explicitly define the expected value so that it's clear what you're comparing against.

File: skbio/stats/distance/tests/test_base.py [hunk 15]
```diff
@@ -467,6 +468,32 @@ def test_to_data_frame_default_ids(self):
...
+        assert_series_almost_equal(series, exp)
+
+    def test_to_series_3x3(self):
```
Comment (jairideout) [inline skbio/stats/distance/tests/test_base.py]: Can you change this test to use a 4x4 matrix, e.g. the one I defined [here](https://github.com/biocore/scikit-bio/pull/1406/files#r73383377)? A matrix larger than 3x3 is useful here to test the ordering of ID pairs in the series.

Can you also add a test case for a 2x2 matrix? That way we'll have test cases for 1x1 (empty series), 2x2 (single value series), and 4x4 (multi-valued series).

File: skbio/stats/distance/tests/test_base.py [hunk 16]
```diff
@@ -467,6 +468,32 @@ def test_to_data_frame_default_ids(self):
...
+        series = self.dm_3x3.to_series()
+        exp = pd.Series([0.0, 0.01, 4.2, 0.01, 0.0, 12.0, 4.2, 12.0, 0.0],
```
Comment (jairideout) [inline skbio/stats/distance/tests/test_base.py]: It's shorter and more readable to define the expected series like this (assuming a 3x3 example matrix):

```python
exp = pd.Series([1.0, 5.0, 42.0], index=[('a', 'b'), ('a', 'c'), ('b', 'c')])
```

Pandas will convert the list of tuples into a `pd.MultiIndex`.

**Comment by jairideout** [comment]
@virginiasaulnier I'm done reviewing. The code, docs, and tests are really nice; most of my comments centered around using a condensed form series, which requires the method to exist on `DistanceMatrix` but not `DissimilarityMatrix`.

Let me know if you have any questions about the requested changes and thanks again for working on this!

File: skbio/stats/distance/tests/test_base.py (hunk 16 above)
Comment (virginiasaulnier) [inline skbio/stats/distance/tests/test_base.py]: My new tests don't seem to be reading the index in the way you suggested.  The assert_series_almost_equal is expecting the MultiIndex levels and labels.  I may have implemented incorrectly.

Comment (jairideout) [inline skbio/stats/distance/tests/test_base.py]: The expected values in the tests are incorrect (they're including both triangles), so those will need to be updated. I also gave you the wrong code, this should work:

```python
exp = pd.Series([0.01, 4.2, 12.0], index=pd.Index([('a', 'b'), ('a', 'c'), ('b', 'c')]))
```

**Comment by jairideout** [comment]
@virginiasaulnier I believe I've answered your question about the failing tests [here](https://github.com/biocore/scikit-bio/pull/1406#discussion_r74350477), please let me know if you have any further questions and when this is ready for review again.

**Comment by jairideout** [comment]
Thanks for your work on this @virginiasaulnier! I'm closing this in favor of https://github.com/biocore/scikit-bio/pull/1453, which contains your work here (I addressed my comments from the previous review).
